tests/p2p/test_sharding.py

```diff
@@ -145,75 +130,47 @@ class InvalidShardingPeer(ShardingPeer):
-        address = Address("127.0.0.1", port, port)
-
-        node = Node(private_key.public_key, address)
-
-        server = ShardingServer(
-            private_key,
-            port,
-            chain=None,
-            chaindb=None,
-            headerdb=None,
-            base_db=None,
-            network_id=9324090483,
-            min_peers=0,
-            peer_class=ShardingPeer
-        )
-        monkeypatch.setattr(server, 'add_nat_portmap', mock_coroutine)
-        monkeypatch.setattr(server, '_discover_upnp_device', mock_coroutine)
-        asyncio.ensure_future(server.run())
-        # Give the server a chance to start all its listeners and the syncer.
-        await asyncio.sleep(1)
-
-        peer_tuples.append(PeerTuple(
-            node=node,
-            server=server,
-            syncer=server.syncer,
-        ))
-
-    # connect peers to each other
-    await asyncio.gather(*[
-        peer_tuples[i].server.peer_pool._connect_to_nodes([peer_tuples[j].node])
-        for i, j in connections
-    ])
-    for i, j in connections:
-        peer_remotes = [peer.remote for peer in peer_tuples[i].server.peer_pool.peers]
-        assert peer_tuples[j].node in peer_remotes
+async def test_shard_syncer(connections, request, event_loop):
+    peers_by_server = {}
+    for server_id1, server_id2 in connections:
+        peer1, peer2 = await get_directly_linked_sharding_peers(request, event_loop)
+        peers_by_server.setdefault(server_id1, []).append(peer1)
+        peers_by_server.setdefault(server_id2, []).append(peer2)
+
+    syncers = []
+    for _, peers in sorted(peers_by_server.items()):
+        peer_pool = MockPeerPoolWithConnectedPeers(peers)
+        shard_db = ShardDB(MemoryDB())
+        syncer = ShardSyncer(Shard(shard_db, 0), peer_pool)
+        syncers.append(syncer)
+        asyncio.ensure_future(syncer.run())
 
     # let each node propose and check that collation appears at all other nodes
-    for proposer in peer_tuples:
-        collation = proposer.syncer.propose()
+    for proposer in syncers:
+        collation = proposer.propose()
         await asyncio.wait_for(asyncio.gather(*[
-            peer_tuple.syncer.collations_received_event.wait()
-            for peer_tuple in peer_tuples
-            if peer_tuple != proposer
-        ]), timeout=10)
-        for peer_tuple in peer_tuples:
-            assert peer_tuple.syncer.shard.get_collation_by_hash(collation.hash) == collation
+            syncer.collations_received_event.wait()
+            for syncer in syncers
+            if syncer != proposer
+        ]), timeout=2)
+        for syncer in syncers:
+            assert syncer.shard.get_collation_by_hash(collation.hash) == collation
 
     # stop everything
-    cancel_token.trigger()
-    await asyncio.gather(*[peer_tuple.server.cancel() for peer_tuple in peer_tuples])
-    await asyncio.gather(*[peer_tuple.syncer.cancel() for peer_tuple in peer_tuples])
+    all_peers = itertools.chain(
```

Comment:
should this be in a `finally` block to ensure it runs?

Reply:
Actually, the one for peers is unnecessary as `get_directly_linked_peers()` already sets up a pytest finalizer that does that. I'll change the one that cancels syncers to use a finalizer as well